### PR TITLE
7.0 issue 3492

### LINKED ISSUE #3492: [8.0] cannot cancel PO containing only service product and with invoicing control = 'Based on incoming shipments'
Impacted versions:
* 8.0

Steps to reproduce:
1. create a purchase order only with service product(s) and Invoicing Control = 'Based on incoming shipments'
2. press 'Cancel' button

Current behavior:
* nothing happens (the PO status remains as 'Purchase Confirmed')

Expected behavior:
* PO status is changed to 'Cancelled'

User should not be processing PO with Invoicing Control 'Based on incoming shipments' if it only contains service products.  However, the problem is that there is no way to fix the PO if you forget to correct the Invoicing Control.

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Fixes #3492

### CODE REVIEW

File: addons/purchase/purchase.py [hunk 1]
```diff
@@ -149,6 +149,20 @@ def _get_journal(self, cr, uid, context=None):
...
+            for order_line in order.order_line:
+                if (
+                    not order_line.product_id
+                    or
+                    order_line.product_id
+                    and order_line.product_id.type not in ('product',
+                                                           'consu')
+                ):
+                    res[order.id] = True
+        return res
```
Comment (jco-odoo) [inline addons/purchase/purchase.py]: I don't agree.  It should be if not(all([x.product_id.type not in ('product', 'consu') for x in order.order_line]), so has_only_non_stockable_item

**Comment by jbeficent** [comment]
@jco-odoo The basic issue now is that you cannot mix stockable and non-stockable items in a PO, and then select the invoicing control based "Based on incoming shipments". Because then you cannot invoice the non-stockable lines afterwards.

What would you do in this situation where it's mixed? You have to restrict the user from doing so (not letting them set this invoicing control), if you are not supporting him afterwards. That's why the proposal is to restrict the user from choosing this method if there's at least one non-stockable item.





But you also have to consider the scenario where no product is added to the line. In that case the invoicing control 'Based on incoming shipments' does not make sense either.

**Comment by jco-odoo** [comment]
@jbeficent When you put a service on a purchase order among stockable products and you invoice the picking, it will also put the service on the invoice.  (if it has not been invoiced already)

**Comment by jco-odoo** [comment]
Ok, but probably not in v7...

**Comment by dmo-odoo** [comment]
Merged in https://github.com/odoo/odoo/commit/a81a7513b2c84408cc376838223d090e3a46132a

**Comment by jbeficent** [comment]
Hi @dmo-odoo. I can see that you just commited this fix. But @jco-odoo did not agree.

After reviewing the scenario in runbot with 7.0, I agree with @jco-odoo. My fix is too restrictive, because it prevents someone from confirming a PO that contains stockable and service items at the same time. But I just tested that in this scenario it is now possible to cancel the PO.

I will submit a new PR.

**Comment by dmo-odoo** [comment]
Hello,

I discussed with jco by mail and he told me the fix was good for v7 but not for v8, I assumed his last comment should have read "Ok, but probably not in v8..." instead of v7.

I'm working on a fix for v8 which implements the behavior he requested, see https://github.com/odoo-dev/odoo/commit/e86e3fdb5d44d9022d0bcdee65dba095049f0e4c.

Best regards,
David

**Comment by dmo-odoo** [comment]
After a long discussion with jco, we decided to adopt the same behavior in 7.0 and 8.0:
https://github.com/odoo/odoo/commit/6758b4cfc5ab5a9bf54c30aa14ed2dfd2fbf534d

**Pull request closed without merging**